demos/common/utilities/sample_adu_jws_mbedtls.c

```diff
@@ -0,0 +1,796 @@
+                                             pucN, ulNLength,
+                                             NULL, 0,
+                                             NULL, 0,
+                                             NULL, 0,
+                                             pucE, ulELength );
+
+    if( lMbedTLSResult != 0 )
+    {
+        LogError( ( "[JWS] mbedtls_rsa_import_raw res: %i", lMbedTLSResult ) );
+        mbedtls_rsa_free( &ctx );
+        return lMbedTLSResult;
+    }
+
+    lMbedTLSResult = mbedtls_rsa_complete( &ctx );
+
+    if( lMbedTLSResult != 0 )
+    {
+        LogError( ( "[JWS] mbedtls_rsa_complete res: %i", lMbedTLSResult ) );
+        mbedtls_rsa_free( &ctx );
+        return lMbedTLSResult;
+    }
+
+    lMbedTLSResult = mbedtls_rsa_check_pubkey( &ctx );
+
+    if( lMbedTLSResult != 0 )
+    {
+        LogError( ( "[JWS] mbedtls_rsa_check_pubkey res: %i", lMbedTLSResult ) );
+        mbedtls_rsa_free( &ctx );
+        return lMbedTLSResult;
+    }
+
+    /* RSA */
+    lMbedTLSResult = mbedtls_rsa_pkcs1_decrypt( &ctx, NULL, NULL, MBEDTLS_RSA_PUBLIC, &ulDecryptedLength, pucSignature, pucBuffer, jwsRSA3072_SIZE );
+
+    if( lMbedTLSResult != 0 )
+    {
+        LogError( ( "[JWS] mbedtls_rsa_pkcs1_decrypt res: %i", lMbedTLSResult ) );
+        mbedtls_rsa_free( &ctx );
+        return lMbedTLSResult;
+    }
+
+    mbedtls_rsa_free( &ctx );
+
+    xResult = prvJWS_SHA256Calculate( pucInput, ulInputLength,
+                                      pucShaBuffer );
+
+    /* TODO: remove this once we have a valid PKCS7 parser. */
+    shaMatchResult = memcmp( pucBuffer + jwsPKCS7_PAYLOAD_OFFSET, pucShaBuffer, jwsSHA256_SIZE );
+
+    if( shaMatchResult )
+    {
+        LogError( ( "[JWS] SHA of JWK does NOT match" ) );
+        xResult = 1;
+    }
+
+    return xResult;
+}
+
+static uint32_t prvFindJWSValue( AzureIoTJSONReader_t * pxPayload,
+                                 az_span * pxJWSValue )
+{
+    AzureIoTResult_t xResult = eAzureIoTSuccess;
+    AzureIoTJSONTokenType_t xJSONTokenType;
+
+    azureiotresultRETURN_IF_FAILED( AzureIoTJSONReader_NextToken( pxPayload ) );
+
+    while( xResult == eAzureIoTSuccess )
+    {
+        if( AzureIoTJSONReader_TokenIsTextEqual( pxPayload, ( const uint8_t * ) jws_sjwk_json_value, sizeof( jws_sjwk_json_value ) - 1 ) )
+        {
+            /* Found name, move to value */
+            azureiotresultRETURN_IF_FAILED( AzureIoTJSONReader_NextToken( pxPayload ) );
+            break;
+        }
+        else
+        {
+            azureiotresultRETURN_IF_FAILED( AzureIoTJSONReader_NextToken( pxPayload ) );
+            azureiotresultRETURN_IF_FAILED( AzureIoTJSONReader_SkipChildren( pxPayload ) );
+            xResult = AzureIoTJSONReader_NextToken( pxPayload );
+        }
+    }
+
+    azureiotresultRETURN_IF_FAILED( AzureIoTJSONReader_TokenType( pxPayload, &xJSONTokenType ) );
+
+    if( ( xResult != eAzureIoTSuccess ) && ( xJSONTokenType == eAzureIoTJSONTokenSTRING ) )
```

Comment:
I don't understand this line.  (some further comments in a response below...)

First, if `xResult` isn't SUCCESS then you cannot trust the value of `xJSONTokenType` because the callee `AzureIoTJSONReader_TokenType` is not guaranteed to have set it.  So you'd be checking against uninitialized value of it on the stack.  I'm not looking up in spec but isn't this also supposed to be a string?

Naively (see part about me not reading spec) is this

```c
if( ( xResult != eAzureIoTSuccess ) || ( xJSONTokenType != eAzureIoTJSONTokenSTRING ) )
```
==

More feedback - we're checking `azureiotresultRETURN_IF_FAILED( AzureIoTJSONReader_TokenType( pxPayload, &xJSONTokenType ) );` before we're checking the `xResult`.  Since we don't know the state of the pxPayload since we could be in an error state, I think it's safer to just check the `xResult` on its own as a separate line 1st thing out of getting out of this loop.

Reply:
Similiar questions around `prvFindSHA`.